### ISSUE TITLE
[pimatic-ping] multiple IP per Hostname doesn't work

### ISSUE THREAD

**Issue by georg90** [pimatic/pimatic#724]
Hello,
like @mwittig proposed here: http://forum.pimatic.org/topic/261/implemented-pimatic-ping-with-dns/6
I have the same issue, where my laptop has two IP entries for one DNS record:
```
nslookup dell-xps.fritz.box

Name:	dell-xps.fritz.box
Address: 192.168.0.144
Name:	dell-xps.fritz.box
Address: 192.168.0.120
```

Like @sweetpi said only the first IP entry is processed. But since I use pimatic to "wake" my docking station when my laptop becomes available, the first IP address will always be the wifi interface and the second the LAN interface. If connected to the docking station wifi is disabled.

How can this be achieved? Even if ony one IP echos --> device is present

**Comment by sweetpi** [comment]
We would need to change the ping plugin. Currently it uses [dns.lookup](https://github.com/pimatic/pimatic-ping/blob/master/ping.coffee#L54), which only returns the first ip record.

We would need to use [dns.resolve4](https://nodejs.org/api/dns.html#dns_dns_resolve4_hostname_callback) which returns all addresses and then try to ping all of them.

**Comment by georg90** [comment]
Doesn't it say ```'A' (IPV4 addresses, default)``` in the docs? Where addresses are plural ?

**Comment by mwittig** [comment]
No. dns.lookup only return the first address - see first sentence at https://nodejs.org/api/dns.html#dns_dns_lookup_hostname_options_callback

So, we need to use dns.resolve4 instead as @sweetpi pointed out

**Comment by georg90** [comment]
@mwittig Did you have some time to come up with a solution?

**Comment by mwittig** [comment]
@georg90 OK, I'll look into this

**Comment by georg90** [comment]
No hurry ! And many thanks in advance.

**Comment by mwittig** [comment]
@georg90 the feature has been released

**Comment by georg90** [comment]
Thank you @mwittig it seems to work just fine!
Great work & thanks again for taking care.

**Comment by mwittig** [comment]
Thanks. Unfortunately it contains a bug as IP resolve4 pass-through IP addresses as dns.lookup does. Anothe PR is on the way.

**Comment by mwittig** [comment]
@georg90 @sweetpi Due to the switching from `dns.lookup` to `dns.resolve`, the plugin currently can only resolve hostnames to IPv4 addresses. My understanding is, that `dns.lookup` resolves a hostname (e.g. 'google.com') into the first found A (IPv4) or AAAA (IPv6) record if the `family` property is undefined (or no options have been passed in). In contrast to this `dns.resolve` only supports either IPv4 or IPv6 lookups. You cannot do both with a single query. If my findings are correct, this implies a limitation for users in a true IPv6 environment  (well, not many these days, but the code should be future proof). Possible solutions:
 1.  Perform a resolve6 (AAAA) query first, if this fails perform a resolve4 (A) query. This appears to be the default behaviour of the DNS resolver on Linux systems these days.
 2. Add a property "protocolVersion" to the PimaticPresence configuration which specifies whether a resolve6 or resolve4 query shall be performed.
 
 What do you think?

**Comment by georg90** [comment]
I would not complicate things for the user with adding a new property.
As you proposed in option 1 - you could simply try v6 first. Will this generate a lot new traffic or slow down (timeout?) the ping at all?

**Comment by mwittig** [comment]
@georg90 Thanks for your feedback.

> Will this generate a lot new traffic or slow down (timeout?) the ping at all?

The traffic overhead is very low as DNS is a lightweight protocol. A slow down may occur if the DNS server is unreachable for some reason. As the plugin code is asynchronous it won't block and we have take measure against resource exhaustion. 

> I would not complicate things for the user with adding a new property.

I have researched this is a little bit further. The cascaded operation I had in mind with option 1) won't work in all cases. For example, if I try with "google.de" the DNS will return an IPv6 address, but I can't ping it as my router does not support IPv6 (Network is unreachable). If I ping a local IPv6 address (::1) this will works fine. So, IPv6 is enabled on my raspi (which wasn't the case by default - I had to modprobe the ipv6 module). As a variation of option 1) the following can be done: Always perform both, resolve4 and resolve6 and then ping the collected results from both queries. This has the disadvantage, however, that more ping requests have to be performed as all IPv4 and IPv6 adresses will be pinged. Unfortunately, Promise.some won't cancel pending operations if the minimum number of operations succeeded.

> Or use Option 2 with the default set to v4 

This is my  favorite right now as it won't harm most users anywaay as they are running on IPv4.
